rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -40,12 +42,29 @@ private class InMemHotStore[F[_]: Monad, C, P, A, K](
       continuations <- getContinuations(channels)
       _             <- S.modify(cache => cache.copy(cache.continuations.updated(channels, wc :: continuations)))
     } yield ()
+
+  def getData(channel: C): F[List[Datum[A]]] =
+    for {
+      cache <- S.get
+      res <- cache.data.get(channel) match {
+              case None       => HR.getData(channel)
+              case Some(data) => Applicative[F].pure(data)
+            }
+      updatedCache = cache.copy(data = cache.data + (channel -> res))
+      _            <- S.set(updatedCache)
```

Comment:
this 'get' 'set' mechanism won't work if the monad state is shared. and RSpace is shared between many rholang threads. (hence the double lock)

Reply:
Good catch. Due to the double lock it is safe to assume that no two threads are going to modify the data under the same channels. I'm going to add a `TrieMap` holding a cache instance per a list of channels

Reply:
BTW, it depends on the underlying `MonadState` implementation. If, as it is currently done in tests, it is an `MVar` it is going to work